File: Fast-Walsh-Hadamard.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Fast Walsh-Hadamard Transform
# Luis Villasenor
# lvillasen@gmail.com
# 2/27/2017 

import numpy as np
import numba

n_qbits = 16
A = np.zeros(2**n_qbits,dtype=np.complex_) # Set initial value of array A
A[0] =1

@numba.autojit
def set_bit(value, bit):
    return value | (1<<bit)

@numba.autojit
def clear_bit(value, bit):
    return value & ~(1<<bit)

@numba.autojit
def Hn(n_qbits,A):
	isq2 = 1/np.sqrt(2)
	for qbit in range(n_qbits):
		B = np.zeros(2**n_qbits,dtype=np.complex_)
		for j in range(2**n_qbits):
			if A[j] != 0:
				bit_parity=(j>>qbit)%2
				if bit_parity == 0:
					B[j]+=isq2*A[j]
					B[set_bit(j,qbit)]+=isq2*A[j]
				elif bit_parity == 1:
					B[clear_bit(j,qbit)]+=isq2*A[j]
					B[j]+=-isq2*A[j]
		A = B
	return B

C = Hn(n_qbits,A)
print(C) # Transformed array

File: Fast-Walsh-Hadamard.c
#include <iostream>
#include <stdio.h>
#include <math.h>

int main (int argc,char *argv[]) {
   int n_qbits;
   if (argc == 2) {
      n_qbits = atoi(argv[1]);
   }
  int N = pow(2,n_qbits);
  double* A = new double[N]; 
  for (int i = 0 ; i < N; i++) {
        A[i]=0;
      }
  A[0] =1.0; // Initialized to |000..00> 
  double* B = new double[N];  
  printf("Number of qubits: %d\n",n_qbits);
  printf("Number of states: %d\n",N);
  double isq2 = 1./sqrt(2);  
  for (int qbit = 0; qbit < n_qbits; qbit++) {
    for (int i = 0; i < N; ++i) {B[i] = 0;}
    for (int j = 0; j < N; ++j) {
      if (A[j] != 0) {
        int bit_parity=(j>>qbit)%2 ;
        if (bit_parity == 0) {
          B[j] += isq2*A[j];
          int set_bit = j | (1<<qbit);
          B[set_bit] += isq2*A[j];
        }
        else if (bit_parity == 1) {
          B[j] += -isq2*A[j];
          int clear_bit = j & ~(1<<qbit);
          B[clear_bit] += isq2*A[j];
        }
      }
    }
    for (int i = 0; i < N; ++i) {A[i] = B[i];}
  }
  for (int i = 0 ; i < 10; i++) {
    printf("i= %d B[i] %.5f:\n",i,B[i]);
  }
  for (int i = N -10 ; i < N; i++) {
    printf("i= %d  B[i] %.5f:\n",i,B[i]);
  }
  return 0;
}
